### ISSUE TITLE
conflictive expressions in cases statement as integral solution

### ISSUE THREAD

**Issue by mech0m** [sympy/sympy#22066]
The following expression
```python
import sympy as sy

x   = sy.symbols("x")
L   = sy.symbols("L", nonzero = True)
k, l = sy.symbols("k, l",   integer=True, nonnegative = True )

integral = sy.Integral( sy.cos( (x*k*sy.pi) / L) * sy.cos( (x*l*sy.pi) / L), (x, -L, L) )
eq = sy.Eq( integral , sy.simplify(integral.doit()) )
eq
```
results in

![image](https://user-images.githubusercontent.com/38977579/132847655-4b07e791-09ca-4291-b7e2-cc2d7db4fcec.png)

Based on the assumptions given for the symbols the conditions stated in the cases statement are not as expected.
The solution I expected is 
 
![image](https://user-images.githubusercontent.com/38977579/132848017-41c7940d-9a4b-4395-b698-5815b67fb552.png)

There are two things that are bothering me

1. cases statement 0 and 1 are conflictive
2. k = -l is always a false statement as k and l where defined as non negative integers

Is there something I'm missing or is this a bug?

**Comment by oscarbenjamin** [comment]
> * cases statement 0 and 1 are conflictive

What do you mean by "conflictive"? Note that case 1 can only apply if the condition for case 0 is false. Likewise case 2 can only apply if the conditions for case 0 and 1 are both false. So if `k=0` and `l=0` then only case 0 applies.

> 2\. k = -l is always a false statement as k and l where defined as non negative integers

In general it could be true if `k` and `l` are both zero. In the context of the Piecewise it is redundant though because e.g. `k = -l or k = l` is equivalent to just `k = l` given that `k = -l` is only true when `k` and `l` are both zero in which case `k = l` is also true. Also since this is case 1 this can only apply if the condition of case 0 is false so we know that `k` and `l` are not both zero. Also the `(k = 0 and l = 0)` in case 1 can be replaced with false because case 1 only applies when the condition in case 0 is false.

The ideal output then would be:
```python
In [4]: Piecewise((2*L, Eq(k, 0) & Eq(l, 0)), (L, Eq(k, l)), (0, True))
Out[4]: 
⎧2⋅L  for k = 0 ∧ l = 0
⎪                      
⎨ L       for k = l    
⎪                      
⎩ 0       otherwise  
```
which is equivalent to what you showed as expected output if you understand that the cases apply sequentially. The fact that k and l are not both zero in case 1 is implicit. Likewise the "otherwise" case just applies when all the others are True.

Basically if you have:
```
>>> Piecewise((A, B), (C, D), (E, F))
⎧A  for B
⎪        
⎨C  for D
⎪        
⎩E  for F
```
then you should read that as being like the Python code:
```
def f():
    if B:
        return A
    elif D:
        return C
    elif F:
        return E
    else:
        return nan
```

I don't see any "bug" but the expression is somewhat redundant. Further simplification is possible but piecewise simplification needs to be improved.

The PR #17443 does not improve the output in this case but there is work to improve simplification of Piecewise (CC @oscargus).

**Comment by mech0m** [comment]
First of all thanks for your time.

So indeed i was missing that the hierarchy in Piecewise is important. Guess my confusion came from the redundant statements.
However improvement of the Piecewise simplification would be nice from an end user perspective.

How to proceed? Should I create a new issue regarding improvement the output or rename this issue?

**Comment by oscargus** [comment]
It is actually not so much about improving the Piecewise as improving the simplification for relationals. From a Piecewise perspective, the simplification in #17443 will and `not (k = 0 ∧ l = 0)` to the second condition. This should theoretically get rid of the `(k = 0 ∧ l = 0)` as `not (k = 0 ∧ l = 0) and (k = 0 ∧ l = 0) = False`. However, as the same thing is applied to the whole expression we get `(k = l or k = -l) and not (k = 0 ∧ l = 0)` which will be deemed more complex than the original expression.

However, there are some things to take away from there. Like applying `not (k = 0 ∧ l = 0)` to each expression separately if the condition is an or-expression. I'll see if I can incorporate that in #17443.

**Comment by oscargus** [comment]
> Should I create a new issue regarding improvement the output or rename this issue?

I think you can just rename this to reflect simplifications of conditions.

**Comment by smichr** [comment]
Another simplification to consider if, for integer values, the case where you have `Piecewise((e1, i<3),(e2,i<4),(e3,i<5),(e4,True))`; the interior conditions can be changed to `Eq(i,3)` and `Eq(i,4)` and those values of `i` can be substituted into the respective expressions.
